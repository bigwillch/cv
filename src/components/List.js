import React, { useState } from 'react'
import Measure from 'react-measure'
import debounce from 'debounce'
import classNames from 'classnames'

export const List = ({
  children,
}) => {

  const [state, setState] = useState({
    scrollable: false,
    scrolled: false,
  })

  const checkScroll = (measure, contentRect) => {
    measure();
    setState({
      ...state,
      scrolled: contentRect.scroll.left > ((contentRect.scroll.width - contentRect.client.width) / 2) ? true : false,
    })
  }

  const className = classNames(
    state.scrollable && 'scrollable',
    state.scrolled && 'scrolled'
  )

  return (
    <Measure
      client
      scroll
      onResize={(contentRect) => {
        setState({
          ...state,
          scrollable: contentRect.client.width < contentRect.scroll.width ? true : false,
        })
      }}
    >
      {({ measure, measureRef, contentRect }) =>
        <React.Fragment>
          <ul
            ref={measureRef}
            className={className}
            onScroll={state.scrollable
              ? () => debounce(checkScroll(measure, contentRect))
              : null
            }
          >
            {children}
          </ul>
          {state.scrollable &&
            <span>scroll me</span>
          }
        </React.Fragment>
      }
    </Measure>
  )
}
