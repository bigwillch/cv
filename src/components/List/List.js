import React, { useState } from 'react';
import Measure from 'react-measure';
import debounce from 'debounce';
import classNames from 'classnames';

export const List = ({
  children,
}) => {
  const [scrollable, setScrollable] = useState(false);

  const [scrolled, setScrolled] = useState(false);

  const checkScroll = (measure, contentRect) => {
    measure();
    setScrolled(contentRect.scroll.left > ((contentRect.scroll.width - contentRect.client.width) / 2));
  };

  const className = classNames(
    scrollable && 'scrollable',
    scrolled && 'scrolled',
  );

  if (navigator.userAgent === 'ReactSnap') {
    return (
      <ul>
        {children}
      </ul>
    );
  }

  return (
    <Measure
      client
      scroll
      onResize={(contentRect) => setScrollable(contentRect.client.width < contentRect.scroll.width)}
    >
      {({ measure, measureRef, contentRect }) => (
        <>
          <ul
            ref={measureRef}
            className={className}
            onScroll={scrollable
              ? () => debounce(checkScroll(measure, contentRect))
              : null}
          >
            {children}
          </ul>
          {scrollable
            && <span>scroll me</span>}
        </>
      )}
    </Measure>
  );
};
