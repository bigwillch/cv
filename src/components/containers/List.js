import React from 'react'
import Measure from 'react-measure'
import debounce from 'debounce'
import classNames from 'classnames'

export default class List extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      scrollable: false,
      scrolled: false,
      scrolledAmount: 0
    };
  }

  checkScroll = (measure, contentRect) => {
    measure()
    this.setState({
      scrolled: contentRect.scroll.left > ((contentRect.scroll.width - contentRect.client.width) / 2) ? true : false
    }) 
  }

  render() {
    // const scrollable = this.state.clientWidth < this.state.scrollWidth ? true : false
    const className = classNames(
      this.state.scrollable && 'scrollable',
      this.state.scrolled && 'scrolled'
    )

    return (
      <Measure
        client
        scroll
        onResize={(contentRect) => {
          this.setState({
            scrollable: contentRect.client.width < contentRect.scroll.width ? true : false,
          })
        }}
      >
        {({ measure, measureRef, contentRect }) =>
          <React.Fragment>
            <ul 
              ref={measureRef} 
              className={className}
              onScroll={this.state.scrollable ? (e) => debounce(this.checkScroll(measure, contentRect)) : null }
            >
              {this.props.children}
            </ul>
            { this.state.scrollable &&
              <span>scroll me</span>
            }
          </React.Fragment>
        }
      </Measure>
    )
  }
}
