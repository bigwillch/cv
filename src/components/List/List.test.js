import React from 'react';
import * as Measure from 'react-measure';
// const Measure = jest.requireActual('react-measure');
import toJson from 'enzyme-to-json';
import { clear, mockUserAgent } from 'jest-useragent-mock'
import { List } from './List';

// jest.mock('react-measure');

// jest.mock('react-measure', () => () => <span>MultiSelect</span>)

describe('List', () => {

  // beforeEach(() => {
  //   jest.resetModules();
  // });

  afterEach(() => {
    clear();
  })

  const ExampleList = (
    <List>
      <li>Item 1</li>
      <li>Item 2</li>
      <li>Item 3</li>
    </List>
  );

  it('should just return a list if pre-rendered via React-Snap', () => {
    mockUserAgent('ReactSnap')
    const wrapper = mount(ExampleList);
    expect(wrapper.first().childAt(0).equals(ExampleList));
  });

  it('should be wrapped in Measure if not pre-rendered', () => {
    const wrapper = mount(ExampleList);
    expect(wrapper.first().childAt(0).type()).toEqual(Measure.default);
    expect(wrapper.find(Measure).find(ExampleList));
  });

  it('should be scrollable if viewport is narrow', () => {

    const contentRect = {
      scroll: {
        left: 0,
        width: 200,
      },
      client: {
        width: 100,
      }
    };

    const measureSpy = jest
      .spyOn(Measure, 'default')
      .mockImplementation(
        ({ children }) => (
          <>
          {
            children({
              measure: jest.fn(),
              measureRef: null,
              contentRect,
            })
          }
          </>
        )
      );

    const wrapper = mount(ExampleList);
    const onResize = wrapper.find(measureSpy).prop('onResize');
    onResize(contentRect);
    wrapper.update();

    expect(toJson(wrapper)).toMatchSnapshot();

    measureSpy.mockRestore();
  });
});
