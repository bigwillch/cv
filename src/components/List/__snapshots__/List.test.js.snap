// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should be scrollable if viewport is narrow 1`] = `
<List>
  <mockConstructor
    client={true}
    onResize={[Function]}
    scroll={true}
  >
    <ul
      className="scrollable"
      onScroll={[Function]}
    >
      <li>
        Item 1
      </li>
      <li>
        Item 2
      </li>
      <li>
        Item 3
      </li>
    </ul>
    <span>
      scroll me
    </span>
  </mockConstructor>
</List>
`;
